### PR TITLE
xen-api-client: depend on xapi-types and xapi-client and drop the need to autogenerate the bindings for xen-api

### PR DESCRIPTION
**Pull request by mseri**
Depends on:
- https://github.com/xapi-project/xen-api/pull/3274
- https://github.com/xapi-project/xs-opam/pull/153

Please **don't merge** until what follows has been discussed and resolved.

This removes the need to autogenerate the bindings for xen-api. This requires us verisoning `xapi-types` and `xapi-client` and move them to `xs-opam/xs` in order to make sure that we know which bindings are generated.

### CODE REVIEW

**Comment by lindig** [comment]
This is quite a big change and I'm not sure I understand the bigger picture. Could you explain either as a comment or in the commit message, what this is about?

**Comment by mseri** [comment]
You can ignore the `Fix all build warnings`, this brings no functional change.

Until now, we were relying on a manual process (only automated in the last PR) to generate the api bindings for xapi. Practically this meant: build xapi, generate the `aPI.ml` and `client.ml` files, copy a few other files from the repository, and drop them in the lib folder with some minor cosmetic changes. Then commit that the API was updated and release a new version. 

This means that each time we changed something in xapi we have a long error-prone process of updating the api, and that `xen-api-client` versions works only with certain specific versions of xapi.

The new `jbuilder` port of xapi, allow us to get the api bindings (the same we would have had manually) simply requiring the relevant xapi subpackages. In this way xen-api-client become completely independent from xapi and provides bindings for the installed version of `xapi-types` and `xapi-clients`, without requiring any manual intervention any longer.

That is the upside. The downside is that in `xs-opam` we need to make sure that we ship the same `xapi-types` and `xapi-client` versions as what is merged on ring3 or master. We can keep the `.master` one in `xs-extras` that work as an override for dev environment, but we need to have stable releases in `xs` (of just these two packages) that we move forward with the versions in the xenserver master branch when having merge to master (I can explain it better to you and @robhoes tomorrow in the office in front of a whiteboard if you want).

File: async_examples/list_vms.ml [hunk 1]
```diff
@@ -25,19 +24,19 @@ let password = ref "password"
...
   let rpc = make !uri in
-  Session.login_with_password rpc !username !password "1.0" "list_vms"
+  Session.login_with_password ~rpc ~uname:!username ~pwd:!password ~version:"1.0" ~originator:"list_vms"
   >>= fun session_id ->
-  VM.get_all_records rpc session_id
+  VM.get_all_records ~rpc ~session_id
   >>= fun vms ->
   List.iter
-    ~f:(fun (vm, vm_rec) ->
+    ~f:(fun (_vm, vm_rec) ->
```
Comment (lindig) [inline async_examples/list_vms.ml]: Can we drop the first argument? `~f:(fun (_, vm_rec)`?

Comment (mseri) [reply async_examples/list_vms.ml]: Sure

**Comment by mseri** [comment]
As an example:

- without this PR, after https://github.com/xapi-project/xen-api/pull/3278/files was merge, we have to regenerate the API bindings to make the `xen-api-client` working again with the current master branch of xapi
- with this PR, it is enough to `opam upgrade xapi` to get an updated `xen-api-client` with the correct bindings

**Comment by mseri** [comment]
Note the travis build is now failing because `xapi-types` depends on `xapi-libs-transitional` (requiring xen devel dependencies and, transitively, systemd) and `xapi-datamodel` (that depends on `xenstore` and `xapi-libs-transitional`).

Restructuring `xapi-libs-transtional` more finely or dropping the dependency will remove the need of such amount of unnecessary libraries.

I would say that this is another downside of this PR, although this one touches us less directly (as we already have such dependencies installed anyway)

**Comment by mseri** [comment]
Note the travis build is now failing because `xapi-types` depends on `xapi-libs-transitional` (requiring xen devel dependencies and, transitively, systemd) and `xapi-datamodel` (that depends on `xenstore` and `xapi-libs-transitional`).

Restructuring `xapi-libs-transtional` more finely or dropping the dependency will remove the need of such amount of unnecessary libraries.

I would say that this is another downside of this PR, although this one touches us less directly (as we already have such dependencies installed anyway)

**Comment by lindig** [comment]
This is in team/ring3/master (as part of xs-opam) but not yet in master.

**Pull request merged**